trino/sqlalchemy/datatype.py

```diff
@@ -75,16 +74,58 @@ def __init__(self, precision=None, timezone=False):
 
 
 class JSON(TypeDecorator):
-    impl = String
+    impl = JSON
 
-    def process_bind_param(self, value, dialect):
-        return json.dumps(value)
+    def bind_expression(self, bindvalue):
+        return func.JSON_PARSE(bindvalue)
 
-    def process_result_value(self, value, dialect):
-        return json.loads(value)
 
-    def get_col_spec(self, **kw):
-        return 'JSON'
+class _FormatTypeMixin:
+    def _format_value(self, value):
+        raise NotImplementedError()
+
+    def bind_processor(self, dialect):
+        super_proc = self.string_bind_processor(dialect)
+
+        def process(value):
+            value = self._format_value(value)
+            if super_proc:
+                value = super_proc(value)
+            return value
+
+        return process
+
+    def literal_processor(self, dialect):
+        super_proc = self.string_literal_processor(dialect)
+
+        def process(value):
+            value = self._format_value(value)
+            if super_proc:
+                value = super_proc(value)
+            return value
+
+        return process
+
+
+class JSONIndexType(_FormatTypeMixin, sqltypes.JSON.JSONIndexType):
+    def _format_value(self, value):
+        if isinstance(value, int):
+            value = "$[%s]" % value
+        else:
+            value = '$.%s' % value
+        return value
+
+
+class JSONPathType(_FormatTypeMixin, sqltypes.JSON.JSONPathType):
+    def _format_value(self, value):
+        return "$%s" % (
+            "".join(
+                [
+                    "[%s]" % elem if isinstance(elem, int) else '.%s' % elem
```

Comment:
why not quote with `[` always? e.g. if `elem` has a dash then also you need to quote.

Reply:
also maybe extract a well-named static method for this

Reply:
good catch, changed